File: agent/tools/LocaleTool.js
const fs = require('fs');
const path = require('path');
const BaseTool = require('./BaseTool');

class LocaleTool extends BaseTool {
  async execute() {
    this.log('📝 Creating locale files...');
    
    // Get translation results from agent state
    const state = this.agent.stateManager.getState();
    const translateResults = state.context.translateResults || { translations: {} };
    const targetLanguage = state.targetLanguage || 'es';
    
    this.log(`  🔍 Debug: targetLanguage = ${targetLanguage}`);
    this.log(`  🔍 Debug: translateResults = ${JSON.stringify(translateResults, null, 2)}`);
    
    const results = {
      localeFilesCreated: 0,
      files: {}
    };
    
    // Create locale directory if it doesn't exist
    const localeDir = path.join(process.cwd(), 'locales', targetLanguage);
    this.log(`  🔍 Debug: localeDir = ${localeDir}`);
    
    if (!fs.existsSync(localeDir)) {
      this.log(`  📁 Creating locale directory: ${localeDir}`);
      fs.mkdirSync(localeDir, { recursive: true });
    }
    
    // Create Spanish common.json file
    const spanishFile = path.join(localeDir, 'common.json');
    const spanishTranslations = this.buildCommonTranslations(translateResults.translations);
    
    this.log(`  🔍 Debug: Writing Spanish common.json to: ${spanishFile}`);
    this.log(`  🔍 Debug: Spanish translations: ${JSON.stringify(spanishTranslations, null, 2)}`);
    
    try {
      fs.writeFileSync(spanishFile, JSON.stringify(spanishTranslations, null, 2));
      this.log(`  ✅ Successfully wrote Spanish common.json`);
      results.localeFilesCreated++;
      results.files[spanishFile] = Object.keys(spanishTranslations).length;
    } catch (error) {
      this.log(`  ❌ Failed to write Spanish common.json: ${error.message}`);
      throw error;
    }
    
    // Create English common.json file
    const englishDir = path.join(process.cwd(), 'locales', 'en');
    if (!fs.existsSync(englishDir)) {
      this.log(`  📁 Creating English locale directory: ${englishDir}`);
      fs.mkdirSync(englishDir, { recursive: true });
    }
    
    const englishFile = path.join(englishDir, 'common.json');
    const englishTranslations = this.buildEnglishTranslations(translateResults.translations);
    
    this.log(`  🔍 Debug: Writing English common.json to: ${englishFile}`);
    this.log(`  🔍 Debug: English translations: ${JSON.stringify(englishTranslations, null, 2)}`);
    
    try {
      fs.writeFileSync(englishFile, JSON.stringify(englishTranslations, null, 2));
      this.log(`  ✅ Successfully wrote English common.json`);
      results.localeFilesCreated++;
      results.files[englishFile] = Object.keys(englishTranslations).length;
    } catch (error) {
      this.log(`  ❌ Failed to write English common.json: ${error.message}`);
      throw error;
    }
    
    // Create namespace files if needed
    const namespaceFiles = this.createNamespaceFiles(translateResults.translations, localeDir);
    results.localeFilesCreated += Object.keys(namespaceFiles).length;
    Object.assign(results.files, namespaceFiles);
    
    // Store results in agent state
    this.agent.stateManager.updateContext({ localeResults: results });
    
    this.log(`  🔍 Debug: results.localeFilesCreated = ${results.localeFilesCreated}`);
    this.log(`  🔍 Debug: results.files = ${JSON.stringify(results.files, null, 2)}`);
    
    // Update phase to indicate locale creation is complete
    if (results.localeFilesCreated > 0) {
      this.agent.stateManager.updateState({ phase: 'locale_complete' });
      this.log(`  ✅ Locale phase complete - ${results.localeFilesCreated} files created`);
    } else {
      this.log(`  ⚠️ No locale files were created (localeFilesCreated = ${results.localeFilesCreated})`);
    }
    
    return results;
  }
  
  buildCommonTranslations(translations) {
    this.log(`  🔍 Debug: buildCommonTranslations called with: ${JSON.stringify(translations, null, 2)}`);
    
    const common = {};
    
    // Handle array of translations (from TranslateTool)
    if (Array.isArray(translations)) {
      translations.forEach(translation => {
        if (translation.key && translation.translation) {
          common[translation.key] = translation.translation;
        }
      });
    } else if (translations && typeof translations === 'object') {
      // Handle object format (legacy)
      Object.values(translations).forEach(categoryTranslations => {
        if (Array.isArray(categoryTranslations)) {
          categoryTranslations.forEach(translation => {
            if (translation.key && translation.translation) {
              common[translation.key] = translation.translation;
            }
          });
        }
      });
    }
    
    this.log(`  🔍 Debug: built common translations: ${JSON.stringify(common, null, 2)}`);
    return common;
  }
  
  buildEnglishTranslations(translations) {
    this.log(`  🔍 Debug: buildEnglishTranslations called with: ${JSON.stringify(translations, null, 2)}`);
    
    const english = {};
    
    // Handle array of translations (from TranslateTool)
    if (Array.isArray(translations)) {
      translations.forEach(translation => {
        if (translation.key && translation.text) {
          english[translation.key] = translation.text; // Use original English text
        }
      });
    } else if (translations && typeof translations === 'object') {
      // Handle object format (legacy)
      Object.values(translations).forEach(categoryTranslations => {
        if (Array.isArray(categoryTranslations)) {
          categoryTranslations.forEach(translation => {
            if (translation.key && translation.text) {
              english[translation.key] = translation.text; // Use original English text
            }
          });
        }
      });
    }
    
    this.log(`  🔍 Debug: built English translations: ${JSON.stringify(english, null, 2)}`);
    return english;
  }
  
  createNamespaceFiles(translations, localeDir) {
    const namespaceFiles = {};
    
    // Create separate files for different categories if they have many strings
    Object.entries(translations).forEach(([category, strings]) => {
      if (strings.length > 5) { // Only create separate files for categories with many strings
        const fileName = `${category}.json`;
        const filePath = path.join(localeDir, fileName);
        
        const categoryTranslations = {};
        strings.forEach(translation => {
          categoryTranslations[translation.key] = translation.translation;
        });
        
        fs.writeFileSync(filePath, JSON.stringify(categoryTranslations, null, 2));
        namespaceFiles[filePath] = strings.length;
      }
    });
    
    return namespaceFiles;
  }
}

module.exports = LocaleTool;

File: pythonAgent/tools/locale_tool.py
"""
Locale tool for creating locale files.
"""

import json
from pathlib import Path
from typing import Dict, List, Any
from tools.base_tool import BaseTool


class LocaleTool(BaseTool):
    """Tool for creating locale files."""
    
    async def execute(self) -> dict:
        """Execute the locale tool."""
        self.log('Creating locale files...')
        
        # Get translation results from agent state
        state = self.agent.state_manager.get_state()
        translate_results = state.translate_results or {'translations': []}
        target_language = state.target_language or 'es'
        
        results = {
            'locale_files_created': 0,
            'files': {}
        }
        
        # Create locale directory if it doesn't exist
        locale_dir = Path('locales') / target_language
        if not locale_dir.exists():
            self.log(f'  Creating locale directory: {locale_dir}')
            locale_dir.mkdir(parents=True, exist_ok=True)
        
        # Create Spanish common.json file
        spanish_file = locale_dir / 'common.json'
        spanish_translations = self._build_common_translations(translate_results.get('translations', []))
        
        try:
            spanish_file.write_text(json.dumps(spanish_translations, indent=2, ensure_ascii=False), encoding='utf-8')
            self.log(f'  Successfully wrote Spanish common.json')
            results['locale_files_created'] += 1
            results['files'][str(spanish_file)] = len(spanish_translations)
        except Exception as error:
            self.log(f'  Failed to write Spanish common.json: {error}')
            raise error
        
        # Create English common.json file
        english_dir = Path('locales') / 'en'
        if not english_dir.exists():
            self.log(f'  Creating English locale directory: {english_dir}')
            english_dir.mkdir(parents=True, exist_ok=True)
        
        english_file = english_dir / 'common.json'
        english_translations = self._build_english_translations(translate_results.get('translations', []))
        
        try:
            english_file.write_text(json.dumps(english_translations, indent=2, ensure_ascii=False), encoding='utf-8')
            self.log(f'  Successfully wrote English common.json')
            results['locale_files_created'] += 1
            results['files'][str(english_file)] = len(english_translations)
        except Exception as error:
            self.log(f'  Failed to write English common.json: {error}')
            raise error
        
        # Store results in agent state
        self.agent.state_manager.update_state({'locale_results': results})
        
        # Update phase to indicate locale creation is complete
        if results['locale_files_created'] > 0:
            self.agent.state_manager.update_state({'phase': 'locale_complete'})
            self.log(f'  Locale phase complete - {results["locale_files_created"]} files created')
        else:
            self.log(f'  No locale files were created')
        
        return results
    
    def _build_common_translations(self, translations: List[Dict[str, Any]]) -> Dict[str, str]:
        """Build Spanish translations for common.json."""
        common = {}
        
        # Handle array of translations (from TranslateTool)
        if isinstance(translations, list):
            for translation in translations:
                if translation.get('key') and translation.get('translation'):
                    common[translation['key']] = translation['translation']
        elif isinstance(translations, dict):
            # Handle object format (legacy)
            for category_translations in translations.values():
                if isinstance(category_translations, list):
                    for translation in category_translations:
                        if translation.get('key') and translation.get('translation'):
                            common[translation['key']] = translation['translation']
        
        return common
    
    def _build_english_translations(self, translations: List[Dict[str, Any]]) -> Dict[str, str]:
        """Build English translations for common.json."""
        english = {}
        
        # Handle array of translations (from TranslateTool)
        if isinstance(translations, list):
            for translation in translations:
                if translation.get('key') and translation.get('text'):
                    english[translation['key']] = translation['text']  # Use original English text
        elif isinstance(translations, dict):
            # Handle object format (legacy)
            for category_translations in translations.values():
                if isinstance(category_translations, list):
                    for translation in category_translations:
                        if translation.get('key') and translation.get('text'):
                            english[translation['key']] = translation['text']  # Use original English text
        
        return english
